File: Task/CUSIP/Visual-Basic-.NET/cusip.vb
Module Module1

    Function IsCUSIP(s As String) As Boolean
        If s.Length <> 9 Then
            Return False
        End If

        Dim sum = 0
        For i = 0 To 7
            Dim c = s(i)

            Dim v As Integer
            If "0" <= c AndAlso c <= "9" Then
                v = Asc(c) - 48
            ElseIf "A" <= c AndAlso c <= "Z" Then
                v = Asc(c) - 55 ' Lower case letters are apparently invalid
            ElseIf c = "*" Then
                v = 36
            ElseIf c = "#" Then
                v = 38
            Else
                Return False
            End If

            If i Mod 2 = 1 Then
                v *= 2 ' check if odd as using 0-based indexing
            End If
            sum += v \ 10 + v Mod 10
        Next
        Return Asc(s(8)) - 48 = (10 - (sum Mod 10)) Mod 10
    End Function

    Sub Main()
        Dim candidates As New List(Of String) From {
            "037833100",
            "17275R102",
            "38259P508",
            "594918104",
            "68389X106",
            "68389X105"
        }

        For Each candidate In candidates
            Console.WriteLine("{0} -> {1}", candidate, If(IsCUSIP(candidate), "correct", "incorrect"))
        Next
    End Sub

End Module

File: Task/CUSIP/Zen-C/cusip.c
fn is_cusip(s: string) -> bool {
    if strlen(s) != 9 { return false; }
    let sum = 0;
    for i in 0..8 {
        let c = s[i];
        let v: int;
        match c {
            48..=57 => { v = c - 48; },  // '0' to '9'
            65..=90 => { v = c - 55; },  // 'A' to 'Z'
            '*'     => { v = 36; },
            '@'     => { v = 37; },
            '#'     => { v = 38; },
            _       => { return false; }
        }
        if i % 2 == 1 { v = v * 2; } // check if odd as using 0-based indexing
        sum += (v / 10) +  v % 10;
    }
    return s[8] - 48 == (10 - (sum % 10)) % 10;
}

fn main() {
    let candidates = [
        "037833100",
        "17275R102",
        "38259P508",
        "594918104",
        "68389X106",
        "68389X105"
    ];
    for candidate in candidates {
        let b = is_cusip(candidate) ? "correct" : "incorrect";
        println "{candidate} -> {b}";
    }
}
